Fixes #18317 - Clone user role via API

xprazak2 opened the pull request:
I do not clone description since it may be removed in the future based on output we got from component demos. Let me know if it should be cloned as well.

Commit:
Fixes #18317 - Clone user role via API

ares commented:
I think the atrribute is not planned to be removed. It was more about the inline help for description field I believe.

ares commented:
Tests needs fixing `ArgumentError: Factory not registered: locations`

dLobatog commented:
@xprazak2 API needs a permission - didn't check the rest of the PR yet but works fine for admin users from what I see 😄

ares commented:
@xprazak2 please leave a message when you update the PR, I see the permission is there now

ares commented:
[test foreman]

xprazak2 commented:
[test foreman]

ares commented on app/models/role.rb:
```diff
@@ -145,6 +144,11 @@ def disable_filters_overriding
 
+  def clone
+    self.deep_clone(:except  => [:name, :builtin],
+                    :include => [:locations, :organizations, { :filters => :permissions }])
```
would you mind updating https://github.com/theforeman/foreman/blob/develop/app/controllers/roles_controller.rb#L92 to use this `#clone` as well?

ares reviewed (changes requested):
one nit, otherwise looks good to me, once resolved, I'll test it

xprazak2 commented:
I updated controller to use `clone` method as well.

ares commented on app/controllers/api/v2/roles_controller.rb:
```diff
@@ -51,11 +51,30 @@ def destroy
 
+      api :POST, "/roles/:id/clone", N_("Clone a role")
+      param :id, String, :required => true
+      param :new_name, String, :required => true, :desc => N_('Name of the cloned role')
```
I see param `name` is used few lines below, which one is correct? also could it be enhanced so it accepts the whole role param_group? I'd like to set the organization while cloning it as well as description.

xprazak2 replied:
I changed to `:new_name`, which is correct one. Interestingly enough hammer sends both `:name` and `:new_name` with the same value, that is why I haven't discovered it:

![](http://i.imgur.com/Gsq0OIF.png)

I can see the value of adding different description to a newly cloned role, but how setting organizations should work is not obvious. Should new clone have the organizations from original when you do not supply any orgs, or should it have none? I believe it would be better to use update for any modifications.

ares replied:
I think all attributes should be consistent. Description is cloned and would not be appended if I specify new one correct/ So I think orgs and locs should behave the same. If you don't specify any, it should have the same as origin. If I specify some (including empty array), it should set this new collection.

ares reviewed (changes requested):
Works well except the fact I can't specify additional attribute values. Please see the inline comment for more details.

ares reviewed (changes requested):
Please note that tests are failing.

xprazak2 commented:
We can now specify taxonomies and description that will be set for the newly created role.

ares commented:
I'll retest but it seems it breaks katello tests, maybe the role cloning is used there somewhere? There are couple of messages like

`ActiveRecord::AssociationTypeMismatch: Taxonomy(#84531320) expected, got NilClass(#8884800)...`

ares reviewed (changes requested):
I'm sorry I still don't understand where `:new_name` is used. It was not allowed in strong params filters so I think it's ignored. 2 of 3 tests in this PR use `:name` only, third one uses `new_name` but also set role hash which contains name so I don't think they test this attribute. My testing result:

![clone](https://cloud.githubusercontent.com/assets/109773/23542969/4436dd06-fff0-11e6-8c5e-f3f7a4412b50.png)

Using name it works just fine. Please let me know what I'm doing wrong.

ares commented:
Another finding is that cloning role that has some taxonomies set does not keep it for new clone (even if I don't specify organization_ids at all)

I'd expect this sets org 1 for the clone and leaves locations untouched
```json
{
  "role": {
    "name": "Manager of Ares23",
    "organization_ids": [1]
  }
}
```

If I'd like to reset locations to empty list, following should be used
```json
{
  "role": {
    "name": "Manager of Ares23",
    "organization_ids": [1],
    "location_ids": []
  }
}
```

Since it does not work this way right now, I think a test for this would be useful.

xprazak2 commented:
I removed the `:new_name` completely, it is a hammer thing and is not needed here. I do not see how I could have broken functional test for subscriptions/index by adding a new method to role. I will revisit it if katello tests remain red. I added tests to cover taxonomy modification cases .

ares reviewed (approved):
Works nicely, :+1:

ares commented:
Thanks @xprazak2, merged as 8813aee5130dc46b8a713864a272054bb7b73f7d
